script/refresh_sitemap

```diff
@@ -0,0 +1,353 @@
+    end
+    fh.write SITEMAP_LEAF_FOOTER
+  end
+end
+
+def write_sitemap_html(name, pages)
+  file = INDEX_FILE.sub("NAME", name)
+  File.open(file, "w") do |fh|
+    fh.write INDEX_HEADER
+    for str, path, time in pages
+      fh.puts "<a href=\"/#{path}\">#{str}</a><br/>"
+    end
+    fh.write INDEX_FOOTER
+  end
+end
+
+def write_index_xml(sitemaps)
+  now = get_current_time
+  file = SITEMAP_FILE.sub("NAME", "index")
+  File.open(file, "w") do |fh|
+    fh.write SITEMAP_ROOT_HEADER
+    for sitemap in sitemaps
+      url  = "<loc>#{SITEMAP_URL.sub("NAME", sitemap)}</loc>"
+      time = "<lastmod>#{now}</lastmod>"
+      fh.puts "<sitemap>#{url}#{time}</sitemap>"
+    end
+    fh.write SITEMAP_ROOT_FOOTER
+  end
+end
+
+def write_index_html(sitemaps)
+  now = get_current_time
+  file = INDEX_FILE.sub("NAME", "index")
+  File.open(file, "w") do |fh|
+    fh.write INDEX_HEADER
+    for sitemap in sitemaps
+      fh.puts "<a href=\"#{sitemap}.html\">#{sitemap}</a><br/>"
+    end
+    fh.write INDEX_FOOTER
+  end
+end
+
+################################################################################
+
+# Times come out of the mysql adapter as Time objects right now.
+# Sitemaps require "yyyy:mm:ddThh:mm:ss+00:00" format.
+# These functions ensure that all times are Strings in that format.
+def parse_mysql_time(time)
+  format_time(time)
+end
+
+def format_time(time)
+  time.utc.strftime "%Y-%m-%dT%H:%M:%S+00:00"
+end
+
+def get_last_update
+  file = SITEMAP_FILE.sub("NAME", "index")
+  format_time(File.mtime(file)) rescue format_time(100.years.ago)
+end
+
+def get_current_time
+  format_time(Time.now)
+end
+
+################################################################################
+
+def update_robots_file
+  allowed_actions = deduce_allowed_actions
+  disallowed_paths = []
+  for controller in get_list_of_controllers
+    actions  = controller_actions(controller)
+    allow    = allowed_actions[controller] || []
+    disallow = actions - allow
+    disallowed_paths += disallow.map {|action| "/#{controller}/#{action}"}
+  end
+  write_robots_file(disallowed_paths)
+end
+
+def deduce_allowed_actions
+  actions = {}
+  for table, type, path in OBJECT_TYPES
+    add_action(actions, path)
+  end
+  for str, path, freq in STATIC_PAGES
+    add_action(actions, path)
+  end
+  return actions
+end
+
+def add_action(actions, path)
+  controller, action = parse_path(path)
+  return if controller.blank?
+  actions[controller] ||= []
+  actions[controller] << action
+end
+
+def parse_path(path)
+  controller, action = path.split("/")
+  action.sub!(/\?.*/, "") if action
+  return controller, action
+end
+
+def get_list_of_controllers
+  # We could use ApplicationController.descendants, but controllers are loaded
+  # lazily, so there won't actually be any loaded yet.  This is safer.
+  result = []
+  wildcard = CONTROLLER_FILE.sub("NAME", "*")
+  for file in Dir.glob(wildcard)
+    result << match[1] if match = file.match(/(\w+)_controller.rb/)
```

Comment:
Wanted to create an example sitemap to check out, so I ran:

$ script/refresh_sitemap 

and got:

script/refresh_sitemap:325:in `block in get_list_of_controllers': undefined local variable or method `match' for main:Object (NameError)
	from script/refresh_sitemap:324:in `each'
	from script/refresh_sitemap:324:in `get_list_of_controllers'
	from script/refresh_sitemap:286:in `update_robots_file'
	from script/refresh_sitemap:351:in `<main>'